File: qpid/java/systests/src/main/java/org/apache/qpid/server/exchange/MandatoryMessageTest.java
/*
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */
package org.apache.qpid.server.exchange;

import junit.framework.TestCase;

import org.apache.log4j.NDC;

import org.apache.qpid.client.AMQNoRouteException;
import org.apache.qpid.client.transport.TransportConnection;
import static org.apache.qpid.server.exchange.MessagingTestConfigProperties.*;
import org.apache.qpid.server.registry.ApplicationRegistry;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import uk.co.thebadgerset.junit.extensions.util.ParsedProperties;
import uk.co.thebadgerset.junit.extensions.util.TestContextProperties;

/**
 * MandatoryMessageTest tests for the desired behaviour of mandatory messages. Mandatory messages are a non-JMS
 * feature. A message may be marked with a mandatory delivery flag, which means that a valid route for the message
 * must exist, when it is sent, or when its transaction is committed in the case of transactional messaging. If this
 * is not the case, the broker should return the message with a NO_CONSUMERS code.
 *
 * <p><table id="crc"><caption>CRC Card</caption>
 * <tr><th> Responsibilities <th> Collaborations
 * <tr><td> Check that a mandatory message is sent succesfully not using transactions when a consumer is connected.
 * <tr><td> Check that a mandatory message is committed succesfully in a transaction when a consumer is connected.
 * <tr><td> Check that a mandatory message results in no route code, not using transactions, when no consumer is
 *          connected.
 * <tr><td> Check that a mandatory message results in no route code, upon transaction commit, when a consumer is
 *          connected.
 * </table>
 */
public class MandatoryMessageTest extends TestCase
{
    /** Used for debugging. */
    private static final Logger log = LoggerFactory.getLogger(MandatoryMessageTest.class);

    /** Used to read the tests configurable properties through. */
    ParsedProperties testProps = TestContextProperties.getInstance(MessagingTestConfigProperties.defaults);

    /** All these tests should have the mandatory flag on. */
    // private boolean mandatoryFlag = testProps.setProperty(IMMEDIATE_PROPNAME, true);
    private boolean mandatoryFlag = testProps.setProperty(MANDATORY_PROPNAME, true);

    /** Check that an mandatory message is sent succesfully not using transactions when a consumer is connected. */
    public void test_QPID_508_MandatoryOkNoTx() throws Exception
    {
        // Ensure transactional sessions are off.
        testProps.setProperty(TRANSACTED_PROPNAME, false);

        // Send one message with no errors.
        ImmediateMessageTest.PublisherReceiverImpl.testNoExceptions(testProps);
    }

    /** Check that an mandatory message is committed succesfully in a transaction when a consumer is connected. */
    public void test_QPID_508_MandatoryOkTx() throws Exception
    {
        // Ensure transactional sessions are off.
        testProps.setProperty(TRANSACTED_PROPNAME, true);

        // Send one message with no errors.
        ImmediateMessageTest.PublisherReceiverImpl.testNoExceptions(testProps);
    }

    /** Check that an mandatory message results in no route code, not using transactions, when no consumer is connected. */
    public void test_QPID_508_MandatoryFailsNoRouteNoTx() throws Exception
    {
        // Ensure transactional sessions are off.
        testProps.setProperty(TRANSACTED_PROPNAME, false);

        // Set up the messaging topology so that only the publishers producer is bound (do not set up the receiver to
        // collect its messages).
        testProps.setProperty(RECEIVER_CONSUMER_BIND_PROPNAME, false);

        // Send one message and get a linked no consumers exception.
        ImmediateMessageTest.PublisherReceiverImpl.testWithAssertions(testProps, AMQNoRouteException.class);
    }

    /** Check that an mandatory message results in no route code, upon transaction commit, when a consumer is connected. */
    public void test_QPID_508_MandatoryFailsNoRouteTx() throws Exception
    {
        // Ensure transactional sessions are on.
        testProps.setProperty(TRANSACTED_PROPNAME, true);

        // Set up the messaging topology so that only the publishers producer is bound (do not set up the receiver to
        // collect its messages).
        testProps.setProperty(RECEIVER_CONSUMER_BIND_PROPNAME, false);

        // Send one message and get a linked no consumers exception.
        ImmediateMessageTest.PublisherReceiverImpl.testWithAssertions(testProps, AMQNoRouteException.class);
    }

    protected void setUp() throws Exception
    {
        NDC.push(getName());

        // Ensure that the in-vm broker is created.
        TransportConnection.createVMBroker(1);
    }

    protected void tearDown() throws Exception
    {
        try
        {
            // Ensure that the in-vm broker is cleaned up so that the next test starts afresh.
            TransportConnection.killVMBroker(1);
            ApplicationRegistry.remove(1);
        }
        finally
        {
            NDC.pop();
        }
    }
}

File: dotnet/Qpid.Integration.Tests/testcases/MandatoryMessageTest.cs
/*
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */
using System;
using System.Threading;
using log4net;
using NUnit.Framework;
using Apache.Qpid.Messaging;

namespace Apache.Qpid.Integration.Tests.testcases
{
    /// <summary>
    /// MandatoryMessageTest checks that messages sent with the 'mandatory' flag, must either be routed to a valid
    /// queue or returned to the sender when no route is available.
    ///
    /// <p><table id="crc"><caption>CRC Card</caption>
    /// <tr><th> Responsibilities <th> Collaborations
    /// <tr><td> Check default exchange returns unroutable mandatory messages.
    /// <tr><td> Check direct exchange returns unroutable mandatory messages.
    /// <tr><td> Check headers exchange returns unroutable mandatory messages.
    /// <tr><td> Check topic exchange returns unroutable mandatory messages.
    /// </table>
    /// </summary>
    [TestFixture, Category("Integration")]
    public class MandatoryMessageTest : BaseMessagingTestFixture
    {
        /// <summary>Used for debugging purposes.</summary>
        private static ILog log = LogManager.GetLogger(typeof(MandatoryMessageTest));

        /// <summary>Defines the maximum time in milliseconds, to wait for redelivery to occurr.</summary>
        public const int TIMEOUT = 1000;

        /// <summary>Condition used to coordinate receipt of redelivery exception to the sending thread.</summary>
        private ManualResetEvent errorEvent;

        /// <summary>Holds the last received error condition, for examination by the tests sending thread.</summary>
        private Exception lastErrorException;
        
        [SetUp]
        public override void Init()
        {
            base.Init();

            errorEvent = new ManualResetEvent(false);
            lastErrorException = null;
            _connection.ExceptionListener = new ExceptionListenerDelegate(OnException);

            _connection.Start();
        }

        [TearDown]
        public override void Shutdown()
        {
            try
            {
                _connection.Stop();
            } 
            finally 
            {
                base.Shutdown();
            }
        }
        
        /// <summary>
        /// Handles all exception conditions on the connection. The error event is notified and the exception recorded as the last seen.
        /// </summary>
        ///
        /// <param name="e">The asynchronous exception on the connection.</param>
        public void OnException(Exception e)
        {
            lastErrorException = e;
            errorEvent.Set();
        }
        
        [Test]
        public void SendUndeliverableMessageOnDefaultExchange()
        {
            SendOne(null);
        }
        
        [Test]
        public void SendUndeliverableMessageOnDirectExchange()
        {
            SendOne(ExchangeNameDefaults.DIRECT);
        }
        
        [Test]
        public void SendUndeliverableMessageOnTopicExchange()
        {
            SendOne(ExchangeNameDefaults.TOPIC);
        }
        
        [Test]
        public void SendUndeliverableMessageOnHeadersExchange()
        {
            SendOne(ExchangeNameDefaults.HEADERS);
        }
        
        /// <summary>
        /// Sends a single message to the specified exchange with the routing key 'unboundkey', marked as mandatory.
        /// A check is performed to assert that a redelivery error is returned from the broker for the message.
        /// </summary>
        ///
        /// <param name="exchangeName">The name of the exchange to send to.</param>
        private void SendOne(string exchangeName)
        {
            log.Debug("private void SendOne(string exchangeName = " + exchangeName + "): called");
            
            // Send a test message to a unbound key on the specified exchange.
            MessagePublisherBuilder builder = _channel.CreatePublisherBuilder()
                .WithRoutingKey("unboundkey")
                .WithMandatory(true);

            if ( exchangeName != null )
            {
                builder.WithExchangeName(exchangeName);
            }

            IMessagePublisher publisher = builder.Create();
            publisher.Send(_channel.CreateTextMessage("Test Message"));
            
            // Wait for up to the timeout for a redelivery exception to be returned.
            errorEvent.WaitOne(TIMEOUT, true);

            // Asserts that a redelivery exception was returned, and is of the correct type.
            Type expectedException = typeof(AMQUndeliveredException);
            Exception ex = lastErrorException;

            Assert.IsNotNull(ex, "No exception was thrown by the test. Expected " + expectedException);
            Assert.IsInstanceOfType(expectedException, ex.InnerException);
        }
    }
}
